### DISCUSSION TITLE
Use Fail2Ban together with xrdp

### DISCUSSION THREAD

**Discussion by hubeny** [neutrinolabs/xrdp#1973] [Q&A] [answered]
Hi,

we have deployed  xrdp server and use it as  the remote desktop server for our users. Everyting works fine, but....
We want to ban ip addresses with fail2ban after several failed attemps to togin via xrdp. 

There are tree log files on ubuntu where goes logs from xrdp (or logs of actions invoked by login attempt to xrdp)
```/var/log/xrdp-sesman.log```
```/var/log/auth.log```
```/var/log/xrdp.log```

The only usable info for fail2ban is in file ```/var/log/auth.log``` example:
`10:16:04 rdplin xrdp-sesman[34403]: pam_unix(xrdp-sesman:auth): authentication failure; logname= uid=0 euid=0 tty=xrdp-sesman ruser= rhost=  user=testuser
`
However, there is no information about rhost (no ip) in the log. We have build latest version xrdp from sources, but without any change in the log. Further we have found  four years opened feature request #392.  Therefore,  we assume, that use fail2ban together with xrdp is not possible now. 

Can anybody give an advice, how to use these two softwares together (xrdp and fail2ban)?

Thank you,

Jan Hubený

**Comment by matt335672** [comment, marked as answer]
Hi @hubeny 

Can you have a look at #1946 and #1947? I was hoping for some feedback from @MikoyChinese on this feature. I'll be happy to work with you to get this doing what you want.

**Reply by hubeny** [reply]
Hi @matt335672,

we simply need to ban IP addresses with too much failed login attemps (either error in user name or in password). As you know Fail2Ban pareses log files line by line and therefore we just need the info about login failure on one line either in ```xrdp-sesman.log``` or ```auth.log``` or ```xrdp.log```. As you wrote in https://github.com/neutrinolabs/xrdp/pull/1947#issuecomment-884797078 it is "something which can change in next release, but for us it is better than nothing". Be sure that we will try it and give you feedback if you implement this feature.

**Reply by matt335672** [reply]
We're on a freeze at the moment, as we've got a release of v0.9.17 in a week or so.

I'm not that familiar with Fail2Ban I'm afraid. I'm aware of the concepts but I've never used it. Input from people like yourself who do use it is invaluable to us.

You say you're building from source. Is there enough information in @MikoyChinese's [post](https://github.com/neutrinolabs/xrdp/pull/1947#issuecomment-884767405) for you to try his suggestion? I can put together a patch for you if that's useful. We should be able to get something ready for you for v0.9.18 (or whatever the next release is).

**Reply by hubeny** [reply]
Thank you for your reply. It will be convenient for me when send me more info. I need to know in which source file and where  (which method or function) in the file I should add  @MikoyChinese solution suggestion. The patch file is not necessary.  If you add this solution to future releases it will be also good.

**Reply by matt335672** [reply]
When v0.9.17 is released, you can make the change in sesman/scp_v0.c at this line:-

https://github.com/neutrinolabs/xrdp/blob/b481351ad7440e0e05f11f2df5214e46171d4d0e/sesman/scp_v0.c#L160

Or, you can fetch the current devel branch and make the change there, if you're happy doing that.

If that gives you enough to work with Fail2Ban we'll get it into v0.9.18.

Does that make sense? Let me know if you need more info.

**Reply by hubeny** [reply]
I tested your suggestion on current devel branch. It works:
```
[20210826-16:22:30] [INFO ] Socket 8: AF_INET connection received from 127.0.0.1 port 46718
[20210826-16:22:32] [ERROR] pam_authenticate failed: Authentication failure
[20210826-16:22:32] [INFO ] Username or password error for user: hubeny from 172.25.4.101:51676 - socket: 12
```
Therefore, please be so nice and add it to the v0.9.18 release.

**Reply by matt335672** [reply]
Have you got this working with Fail2Ban?

Sorry to ask the explicit question, but I can then comment the line appropriately in the source code so that we know there's an external dependency on it. This should address most of the concern I raised with @MikoyChinese

**Reply by hubeny** [reply]
To be honest, I can build xrdp from sources and test the behaviour with wrong username/password - in the ```xrdp-sesman.log``` are the wished lines which can be parsed with fail2ban. So to your last question, it works. However, the devel branch of xrdp builded from sources on Ubuntu 20.04 LTS is not working with corrent username/password. After login we get blank screen. I cannot found the source of this behaviour.

**Reply by hubeny** [reply]
If you are interested, I put on the web corresponding xrdp.log and xrdp-sesman.log here 

https://cloud.soma-eng.com/s/JS5gLaeys83Z2w6

**Reply by matt335672** [reply]
Looking in your sesman log, it looks like you've got a problem starting the session or the X server.

Have a look in these places:-
- ~/.xorgxrdp.10.log will tell us what the X server is doing.
- `sudo journalctl -S -5m` will give you the last 5 minutes of what's happening on the system, so should include your session. Adjust the timeout to suit.

Have you also rebuilt xorgxrdp? the one which comes with 20.04 is not compatible with the latest devel sources.

**Reply by hubeny** [reply]
I have done these steps:
1) build latest devel  xrdp from sources and did make install and reboot
2) buid xorgxrdp did make install and reboot

The ~/.xorgxrdp.10.log  is not even created in this setup. 

I did not unistall the distribution packages before installing the xrdp a nd the xorgxrdp module from sources. I will give another try and upload also the journalctl.

**Reply by matt335672** [reply]
The uninstall might well be it. If you've got a source build and a package build on the same system they will probably pick up on the shared libraries for the installed build.

**Reply by matt335672** [reply]
@hubeny - I'm about to sign off for a long weekend. I've just raised PR #1976 which you might be interested in. Feel free to comment on it.

**Reply by hubeny** [reply]
The simple procedure 
uninstall ubuntu's packages xrdp and xorgxrdp  
reboot; make and make install of xrdp; reboot; make and make install of xorgxrdp

did not help.

What really helps was:
uninstall ubuntu's packages xrdp and xorgxrdp  reboot
and use the xrdp easy install from https://c-nergy.be/blog/?p=15978
here I use the -c -s command line args and apply the patch into the sources downloaded by git. 

Thank you for your help. Now everything works together with fail2ban.

**Comment by SxnSess** [comment]
**hi @hubeny would like to know what settings you made for fail2ban to monitor xrdp.**

**Comment by padenmoss** [comment]
For xrdp and fail2ban on debian Trixie:
/etc/fail2ban/jail.local
```
[xrdp]
enabled = true
port    = 3389
filter  = xrdp
logpath = /var/log/xrdp-sesman.log
maxretry = 3
bantime  = 3600
```

/etc/fail2ban/filter.d/xrdp.conf
```
[Definition]
failregex = ^.*AUTHFAIL: user=\S* ip=:*.*:<HOST> time=\d*$
maxlines = 1
```

**Reply by padenmoss** [reply]
This failregex ignores blank or default IPv6 addresses.
Contact me@padenmoss.com
